### PR TITLE
EIP 1283 - Net gas metering for SSTORE without dirty maps

### PR DESCRIPTION
**Pull request by mmrozek**
Spec: https://eips.ethereum.org/EIPS/eip-1283

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala [hunk 1]
```diff
@@ -533,21 +533,66 @@ case object MSTORE8 extends OpCode(0x53, 2, 0, _.G_verylow) {
...
+    val refund: BigInt = if(isAfterConstantinopleFork(state)){
+      val originalValue = state.originalStorage.load(offset)
+      if(originalValue == currentValue) { // fresh slot
+        if (newValue.isZero)
+          state.config.feeSchedule.R_sclear
+        else 0
+      } else { // dirty slot
+        val clear =
+          if (!newValue.isZero)
+            -state.config.feeSchedule.R_sclear
```
[the remark this answers is not in the document]
Comment (mmrozek) [reply src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: We could import whole state.config.feeSchedule._ but I wanted to be consistent with the rest usages of fee schedules.

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala [hunk 2]
```diff
@@ -533,21 +533,66 @@ case object MSTORE8 extends OpCode(0x53, 2, 0, _.G_verylow) {
...
     state.withStack(stack1).withStorage(updatedStorage).refundGas(refund).step()
   }
 
   protected def varGas[W <: WorldStateProxy[W, S], S <: Storage[S]](state: ProgramState[W, S]): BigInt = {
-    val (Seq(offset, value), _) = state.stack.pop(2)
-    val oldValue = state.storage.load(offset)
-    if (UInt256(oldValue).isZero && !value.isZero) state.config.feeSchedule.G_sset else state.config.feeSchedule.G_sreset
+    val (Seq(offset, newValue), _) = state.stack.pop(2)
+    val currentValue = state.storage.load(offset)
+    if(isAfterConstantinopleFork(state)){
```
Comment (KonradStaniec) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: Maybe create separate functions for pre-consatninople and post-constatninople rules ?

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala [hunk 3]
```diff
@@ -533,21 +533,66 @@ case object MSTORE8 extends OpCode(0x53, 2, 0, _.G_verylow) {
 
 case object SSTORE extends OpCode(0x55, 2, 0, _.G_zero) {
   protected def exec[W <: WorldStateProxy[W, S], S <: Storage[S]](state: ProgramState[W, S]): ProgramState[W, S] = {
-    val (Seq(offset, value), stack1) = state.stack.pop(2)
-    val oldValue = state.storage.load(offset)
-    val refund: BigInt = if (value.isZero && !UInt256(oldValue).isZero) state.config.feeSchedule.R_sclear else 0
-    val updatedStorage = state.storage.store(offset, value)
+    val (Seq(offset, newValue), stack1) = state.stack.pop(2)
+    val currentValue = state.storage.load(offset)
+    val refund: BigInt = if(isAfterConstantinopleFork(state)){
```
Comment (KonradStaniec) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: Maybe create separate functions for pre-consatninople and post-constatninople rules ?

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala (hunk 2 above)
Comment (mmrozek) [reply src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: In my opinion, it doesn't make sense. There is only 1 usage and it is assigned to the value with meaningful name. `if(isAfterConstantinopleFork(state)) afterConstantinopleForkRules() else beforeConstantinopleForkRules()` is to verbose for me.

**Pull request merged**